File: python/course-repo/solution/05-heartbeats-and-checkpointing/src/webhooks/send_bulk.py
"""Fan out N webhook deliveries as Standalone Activities and wait for all."""

import asyncio
import sys
from datetime import timedelta

from temporalio.client import Client

from .activities import deliver_webhook
from .shared import WEBHOOK_RECEIVER_URL, TASK_QUEUE, WebhookDelivery


async def main(count: int = 30) -> None:
    client = await Client.connect("localhost:7233")
    handles = []
    for i in range(count):
        h = await client.start_activity(
            deliver_webhook,
            args=[WebhookDelivery(
                url=WEBHOOK_RECEIVER_URL,
                payload={"event_id": f"bulk_{i:03d}", "type": "bulk_send"},
                event_id=f"bulk_{i:03d}",
            )],
            id=f"bulk-{i:03d}",
            task_queue=TASK_QUEUE,
            start_to_close_timeout=timedelta(seconds=30),
        )
        handles.append(h)
    await asyncio.gather(*(h.result() for h in handles))
    print(f"All {count} deliveries completed.")


if __name__ == "__main__":
    asyncio.run(main(int(sys.argv[1]) if len(sys.argv) > 1 else 30))

File: typescript/course-repo/solution/04-concurrency-and-rate-limits/src/sendBulk.ts
import { Client, Connection } from '@temporalio/client';
import { TASK_QUEUE, WEBHOOK_RECEIVER_URL } from './shared';

async function main(count: number) {
  const connection = await Connection.connect({ address: 'localhost:7233' });
  const client = new Client({ connection });

  const handles = await Promise.all(
    Array.from({ length: count }, (_, i) =>
      client.activity.start("deliverWebhook", {
        args: [{ url: WEBHOOK_RECEIVER_URL, payload: { eventId: `bulk_${String(i).padStart(3, '0')}`, type: 'bulk_send' }, eventId: `bulk_${String(i).padStart(3, '0')}` }],
        id: `bulk-${String(i).padStart(3, '0')}`,
        taskQueue: TASK_QUEUE,
        startToCloseTimeout: '30 seconds',
      })
    )
  );

  await Promise.all(handles.map(h => h.result()));
  console.log(`All ${count} deliveries completed.`);
  await connection.close();
}

const count = parseInt(process.argv[2] ?? '60', 10);
main(count).catch((err) => {
  console.error(err);
  process.exit(1);
});
